### PR TITLE
[WIP] Add Notification Stream

### PR DESCRIPTION
**Pull request by mf-social**
Hi,

This is a long one - sorry!

## TLDR

This PR generates the foundations for having a notifications stream, so that when you receive a notification it is saved to the device and can be viewed, as per most (all?) other notification platforms.

This was suggested in #398 by @robbiet480 , but I cannot see any further discussion on the matter.

It is fully working as-is at this time, but I've stopped at this point as I don't want to go much further without some kind of discussion over direction (see below).

## What's in it so far?

I've added  a database class to store the messages with the following columns in the table:

 - Unique ID
 - Tag (as sent from homeassistant)
 - Message title
 - Message body
 - Image URL (For attachments)
 - Timestamp

I've also included columns ready for future expansion:

 - Whether the message is incoming or outgoing
 - Weather the messagge has been read by the recipient

## What can it do?

 - Pressing the 'app configuration' label in the sidebar now goes directly in to the messaging stream, with a shortcut to the settings now appearing in the top right of the messsage stream activity as a cog icon.
 - The stream displays a friendly message if there are currently no messages saved in the database
 - Any new notifications received will be automatically saved
 - Sending `clear_notification` with a tag also deletes those tagged messages from the database
 - If the message included a link (eg image), tapping on it opens the link in your default browser
 - Long press to select messages in the stream and open the contextual options (Select All, Delete)

## What Needs to be done?

 - Tests - I haven't done any tests yet because I don't want to create tests for things that aren't going to be used, especially as I don't know much more about the project structure yet.  These will follow once the 'concept' has been approved.
 - It took me a while to get my head around the file structure of the project that I needed to know to add this feature, but I'm confused about the rest.  Do you want the Database class in Util?  Do you want a whole new directory for the Message stream and its associated files?  Some direction on this (and some explanations on how the rest of the project is structured) would be welcome, and I might be able to help with some of the other issues too in due course then.
 - Confirm launch vector - currently the link in the homeassistant sidebar says "app configuration" - ideally this would be changed to "Mobile App" and the icon changed to something less settings-y like cellphone-text.  Unless it would be preferred that there be two separate sidebar labels (one for settings and one for messages)?  Personally I think the latter would add more clutter than it is worth.
 - Translations - I've added some new strings to the master strings file - how do I prompt for them to be translated?  Or do I just leave it and wait for someone to come across it in English in their app and submit a translation later?

## Things I intend to add in the future:

 - Tapping a notification should take you to the notification activity, with the Webview in the back-stack
 - Ability to send message back the other way (and possibly make a 'keyboard' with the most recent actionable notification that was sent, a la Telegram)
 - The message stream should update if you're in it and a new message comes through (or if one is deleted via the clear_tag attribute)
 - Manage images in the messages themselves rather than a link (This will need some kind of architecture discussion as they will ideally need saving to the device when the notification is received) - This would also address #609 
 - ??

## How quickly can you do it?

Not very.  The actual coding in this PR probably only took me a few hours in front of the laptop including checks, but the actual amount of time since I first cloned the repo is somewhere around 3 weeks.  My missus and I have 5 young kids between us and I work a full-time shift-work job that isn't in IT, and have other IT commitments outside.  I usually contribute to the forums and things from my phone, and obviously I can't code android studio on it.  I will happily plod along with these things though when I get chance, but doing things like "discord meetings" are really too time intensive for me unfortunately.  If someone on the dev team can give their thoughts on what's been said here I can adjust anything that  needs adjusting, get some tests in and get this bit merged, then start getting on with the other features.

## I don't want to wait for this feature!

Then you'll have to create your own APK from my fork of the repo.

## What does it look like

- tap App Configuration (Hopefully won't be called that in the frontend when merged)

menu_show

 - Opens your message stream (Note that settings can now be accessed from the cog, top right)

empty_messages

 - Come back out and send yourself some messages (currently the view does not update by itself), then go back in

demo_messages

 - Try sending yourself a message that has a tag, check it is in your stream (see Message C above), then send yourself a message to clear that tag - notice that next time you open your stream the tagged message has been deleted

tagged_message_deleted

- Tap on a message with an atachment (currently shown as a link) to view the attachment in your default browser

opens_link_in_browser

 - Long press to select a message in the stream...

selected_messages

 - If you want to select them all, press the Select All button on the contxt menu...

select_all

- And press the Delete button in the context menu to get rid of any you're done with...
empty_messages

### CODE REVIEW

**Comment by mf-social** [comment]
## Can you add feature parity to IOS?

No - I have no clue about xcode/swift, but I'll happily discuss the logic with anyone from the ios team to ensure that going forward both projects are working towards a similar end product.

**Comment by mf-social** [comment]
...The Travis failure appears to be due to the tests, I haven't touched those 😆

**Comment by mf-social** [comment]
I thought I'd sussed the Travis checks, but apparently not 🤷‍♂️

**Comment by mf-social** [comment]
...travis fixed 🙂

**Comment by balloob** [comment]
This looks cool! I don't have the knowledge to review this, but I do want to comment on using "App Configuration" for this and the suggestion that we rename this. Right now it's not possible to rename this, this is a built-in feature of the Home Assistant Frontend to allow apps wrapping the frontend to specify if they have a config screen or not ([docs](https://developers.home-assistant.io/docs/frontend/external-bus#get-external-config)). This is all hardcoded and translated in the frontend itself and is not part of the app.

**Comment by mf-social** [comment]
Thanks :sunglasses:

Indeed, I was hoping maybe someone could rename the label in the frontend?

Or are you saying there could potentially be several 'external apps'? If so, then how does it know which label opens which app?

Maybe some way to specify the name of the label and the icon for each app? I don't understand the front end code at all so I'm clutching at straws on that bit tbh.

**Comment by mf-social** [comment]
To clarify the above now I've looked at the doc you linked to...

A mobile app currently registers

```
{
  hasSettingsScreen: boolean;
}
```

Which generates the app configuration label on the sidebar, tapping on which opens the native-to-app settings activity. 

Now a mobile app in the future may have several native-to-app activities, this PR creates an activity for the message stream. 

So I'm thinking the neatest way to go about accepting the potential for multiple native-to-app activities is that when the boolean is true, the sidebar label reads "Mobile App" and then navigation between any native-to-app activities is handled in the app itself. 

The alternative would be creating a series of booleans for each potential activity 

```
  hasSettingsScreen: boolean;
  hasMessageStream: boolean;
  hasWearOsConfigurationPage: boolean;
  hasSomeOtherThing: boolean;
```

Each one generating its own label in the sidebar, but the downside to this approach is that the homeassistant code would have to be updated every time someone has a bright idea for a new native-to-app activity, and its also going to make your sidebar even more cluttered. 

Given that realistically settings are only accessed once in a blue moon, it seems sensible to me to use the design pattern I'm suggesting, but I also totally understand that I may not be looking at this from the same angle as others so I'm grateful for any other thoughts on how to proceed. 

The only other option I thought about was to place a floating action button on the webview activity that opens the message stream, but the downside of this is that it might get in the way of other controls on the frontend, and also starts getting messy if other native-to-app activities get added in the future. 

So yeah, I'm open to suggestions on how to approach this going forward, but my preference would be the rename of the label to "mobile app" and a minor change to the icon (obviously not now, when this feature is finished) because I think it's more future proof and fits well with the frontend, and basically makes it seamless between the homeassistant based activities in the webview and the native-to-app activities.

**Comment by balloob** [comment]
These settings are used by any mobile app. Not just the official ones. I think for MVP we should stick with App Configuration as label.

**Comment by mf-social** [comment]
I get that - I'm just saying that going forward the label will always launch a mobile app, but may not always launch a configuration activity.

> I think for MVP we should stick with App Configuration as label

Of course :slightly_smiling_face:

I trust that there are no other issues from your point of view with the concept of the stream itself? That being the case I'm just waiting on review and confirmation of the structure, and can then add some tests, get merged, and get started on the other bits as described in my OP.

**Comment by JBassett** [comment]
I'll see if I can take a look at this today.  @balloob I know @robbiet480 had some thoughts about letting Home Assistant Core itself expose a UI around notifications.  I'm ok personally with either direction we take but wanted to make sure we want to go with saving the messages on device rather than expose from Home Assistant Core itself.

**Comment by mf-social** [comment]
From my perspective saving the notifications on device makes more sense in case you want to go back through your messages when you're not connected for whatever reason.  For the same reason I think images should be saved to device automatically as per the bug report/feature request I linked to in the OP.

**Comment by dshokouhi** [comment]
So there are a 2 things to consider here regarding the history of notifications.  If they are saved to the device then that means the following:

- They won't be easily transferred when user gets a new phone or tablet
- Clearing data wipes them away

Both of those points can be solved by having a backup/restore option.

With that said most users are used to things like Telegram and Discord and Slack for notifications and those retain the history.  Personally I opt for HA saving the text and maybe even linking to a `person` or `notify` service to be able to pull the data if you login to a new device.

For me personally I like to see the full history of things.  So I think saving them on the device is good to have but in the long run we want to sync them across devices.  With that said it probably has a bigger undertaking in core to accomplish that but if you look at #398 it was already suggested to pull them from core so you can display them in either platform.

I do like @mf-social idea about viewing notifications offline, that does have some use case to it.

**Comment by mf-social** [comment]
I think the 'limitations' you're saying regarding the messages not going to new devices without a backup strategy is the same for all notification platforms that are communicating direct with an on-device app (rather than something like email where the message is sent to some server and then the app syncs with the server).

I would personally be allowing the standard 'back-up app to your google account' for this app if it was my choice but that's out of the scope of this PR.  It's only a line of code in the manifest and a small xml file defining what should/shouldn't be saved, so I can submit a small PR for that later if it's agreed that's how it should be done.

> For me personally I like to see the full history of things

Currently you don't get anything close to this with the app, once you have dismissed the notification it is gone forever 🙂

**Comment by dshokouhi** [comment]
Your backup solution would solve my concerns actually, I will open a feature request so we can discuss it there.

> Currently you don't get anything close to this with the app, once you have dismissed the notification it is gone forever 🙂

Technically every android device does get the notification log from the settings widget however that is nowhwere near as neat and tidy and awesome as you have constructed here :) Talking this out more I do like this approach more now that I am thinking about it more.

**Comment by tubalainen** [comment]
@mf-social Thank you for all your hard work.

Its because people like you we all get to enjoy the benefits of open source! ❤️ 

Take care of your five year old (got two monsters here my self :) ) - family first. And when you get a break from the unicorn chats we all love the time you can put into improving things.

I am personally really looking forward to having a notification history including the possibility to save attached photos. That would really make this killer app epic!

**Comment by mf-social** [comment]
Ha, you misread, I don't have a 5 year old - we have 5 children - 17, 11, 10, 9 and 8 so it's a bit of a handful - especially at the moment because we're home schooling them all! :laughing:

Thanks for your comments :slightly_smiling_face:

**Comment by manuelo1983** [comment]
Hello, wow this is a great feature. Actually exactly what I am missing now, because if you wipe away the notifications or they are removed by the system, you have no possibility to follow up on something afterwards. So this possibility with logging and history would be extremely super and helpful. I would be happy if this function will be available in the near future.

Thanks and greetings

File: app/src/main/java/io/homeassistant/companion/android/notifications/NotificationsActivity.kt [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+import android.widget.AbsListView
+import android.widget.AdapterView
+import android.widget.ListView
+import androidx.appcompat.app.AppCompatActivity
+import androidx.cursoradapter.widget.SimpleCursorAdapter
+import io.homeassistant.companion.android.R
+import io.homeassistant.companion.android.settings.SettingsActivity
+import kotlinx.android.synthetic.main.activity_notifications.*
+
+class NotificationsActivity : AppCompatActivity() {
```
Comment (JBassett) [inline app/src/main/java/io/homeassistant/companion/android/notifications/NotificationsActivity.kt:21]: Do we want this to be a full activity or just a fragment under the settings?

File: app/src/main/java/io/homeassistant/companion/android/notifications/NotificationsDB.kt [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    @Throws(SQLException::class)
+    fun deleteMessage(s: String) {
+
+        d!!.delete(SQLITE_TABLE_MESSAGE_STREAM, "$KEY_ENTRY_UID = ? ", arrayOf(s))
+    }
+
+    @Throws(SQLException::class)
+    fun clearMessage(s: String) {
+
+        d!!.delete(SQLITE_TABLE_MESSAGE_STREAM, "$KEY_TAG = ? ", arrayOf(s))
```
Comment (JBassett) [inline app/src/main/java/io/homeassistant/companion/android/notifications/NotificationsDB.kt:63]: Quite the clear ;-)

File: app/src/main/java/io/homeassistant/companion/android/notifications/NotificationsDB.kt [hunk 3]
```diff
@@ -0,0 +1,106 @@
+package io.homeassistant.companion.android.notifications
```
Comment (JBassett) [inline app/src/main/java/io/homeassistant/companion/android/notifications/NotificationsDB.kt:1]: This should probably go in data with all the other persisting stuff?

File: app/src/main/java/io/homeassistant/companion/android/notifications/NotificationsActivity.kt [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+
+            val cursor = listView.getItemAtPosition(position) as Cursor
+            val link = cursor.getString(cursor.getColumnIndexOrThrow("image_uri"))
+
+            if (link != null) {
+
+                try {
+                    startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(link)))
+                } catch (e: ActivityNotFoundException) {
+                    e.printStackTrace()
```
Comment (JBassett) [inline app/src/main/java/io/homeassistant/companion/android/notifications/NotificationsActivity.kt:94]: Logger

File: app/src/main/AndroidManifest.xml [hunk 5]
```diff
@@ -94,9 +94,11 @@
             android:supportsPictureInPicture="true"
             android:resizeableActivity="true"
             android:configChanges="orientation|screenSize|smallestScreenSize|screenLayout" />
+        <activity android:name=".notifications.NotificationsActivity"
+            android:parentActivityName=".webview.WebViewActivity" />
         <activity
             android:name=".settings.SettingsActivity"
-            android:parentActivityName=".webview.WebViewActivity" />
+            android:parentActivityName=".notifications.NotificationsActivity" />
```
[the remark this answers is not in the document]
Comment (mf-social) [reply app/src/main/AndroidManifest.xml:101]: I can do it like that if you want, but it seems counterintuitive. I don't know of any other app where you go in to the settings first to open a different activity. In fact, settings should be something that you go to once in a blue moon.

File: app/src/main/java/io/homeassistant/companion/android/notifications/MessagingService.kt [hunk 6]
```diff
@@ -105,6 +106,7 @@ class MessagingService : FirebaseMessagingService() {
         val messageId = tag.hashCode()
 
         notificationManager.cancel(tag, messageId)
+        clearFromDb(tag)
```
[the remark this answers is not in the document]
Comment (mf-social) [reply app/src/main/java/io/homeassistant/companion/android/notifications/MessagingService.kt:109]: What's the point in cancelling the message if you still want it?

File: app/src/main/java/io/homeassistant/companion/android/notifications/MessagingService.kt [hunk 7]
```diff
@@ -154,6 +156,57 @@ class MessagingService : FirebaseMessagingService() {
...
+        val title = data[TITLE]
+        val message = data[MESSAGE]
+        val image = data[IMAGE_URL]
+        val time = DateFormat.getDateTimeInstance().format(System.currentTimeMillis())
+        val read = true // Change this to false when ready to monitor read state
+
+        try {
+
+            db.open()
+            db.addMessage(tag, title, message, image, time, read, "incoming")
```
[the remark this answers is not in the document]
Comment (mf-social) [reply app/src/main/java/io/homeassistant/companion/android/notifications/MessagingService.kt:176]: Sorry, I don't understand this one?

File: app/src/main/java/io/homeassistant/companion/android/notifications/MessagingService.kt [hunk 8]
```diff
@@ -167,6 +220,7 @@ class MessagingService : FirebaseMessagingService() {
             }
         } else {
             WebViewActivity.newInstance(this, url)
+            // TODO: Open Notifications Activity here, with webview primed in back-stack
```
[the remark this answers is not in the document]
Comment (mf-social) [reply app/src/main/java/io/homeassistant/companion/android/notifications/MessagingService.kt:223]: That's fine, I'll just remove the TODO - but generally tapping on a notification for a message takes you to the message stream in all other apps.

File: app/src/main/java/io/homeassistant/companion/android/notifications/NotificationsActivity.kt (hunk 1 above)
Comment (mf-social) [reply app/src/main/java/io/homeassistant/companion/android/notifications/NotificationsActivity.kt:21]: Ha, I don't know, you're the oracle :)

I would prefer to keep settings completely separate from, non-settings on device activities, but it's up to you.

File: app/src/main/java/io/homeassistant/companion/android/notifications/NotificationsActivity.kt [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+            db?.open()
+        } catch (e: Exception) {
+
+            e.printStackTrace()
+        }
+    }
+
+    private fun generateStream() {
+
+        stream = message_stream
```
[the remark this answers is not in the document]
Comment (mf-social) [reply app/src/main/java/io/homeassistant/companion/android/notifications/NotificationsActivity.kt:58]: It might not always be a listview, but it will always be a message stream. Depending on what features we end up giving it it might end up better being a recyclerview, but for initial simplicity I went with listview and simple cursor.

File: app/src/main/java/io/homeassistant/companion/android/notifications/NotificationsActivity.kt [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+
+            if (link != null) {
+
+                try {
+                    startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(link)))
+                } catch (e: ActivityNotFoundException) {
+                    e.printStackTrace()
+                }
+            }
+        }
```
[the remark this answers is not in the document]
Comment (mf-social) [reply app/src/main/java/io/homeassistant/companion/android/notifications/NotificationsActivity.kt:97]: What would they expect instead?

File: app/src/main/java/io/homeassistant/companion/android/notifications/NotificationsDB.kt (hunk 3 above)
Comment (mf-social) [reply app/src/main/java/io/homeassistant/companion/android/notifications/NotificationsDB.kt:1]: Will do.

File: app/src/main/java/io/homeassistant/companion/android/notifications/NotificationsDB.kt (hunk 2 above)
Comment (mf-social) [reply app/src/main/java/io/homeassistant/companion/android/notifications/NotificationsDB.kt:63]: I don't understand this one, sorry?

**Comment by mf-social** [comment]
Where am I logging to?

**Comment by mf-social** [comment]
@JBassett - Will hopefully get chance to do some more work on this at the weekend, but could do with some further clarity as above please.  Also please confirm the testing framework if you want me to use it?

**Comment by manuelo1983** [comment]
Only for my understanding, because this issue is closed now. Is this feature now included in Home Assistant or does this not come at all?

**Comment by JBassett** [comment]
I won't speak for @mf-social but based on the comments and build changes he might just be re-basing his code.  This functionality is not currently implemented in HA Core.

**Comment by dshokouhi** [comment]
Hey @mf-social I know its been a lil bit on this PR and that you are incredibly busy but wanted to reach out and see if you need any help here.  I see that there were still some unanswered questions (i.e. logging and tests), do you still need a response on those?   I know the code base has changed a bit too since this was created as we have setup Room and moved a few things around.  So anything you need to move forward here happy to help 🙏  Again totally understand how busy you are and appreciate your contribution here 😃

**Comment by mf-social** [comment]
Hi @dshokouhi , there were way too many unanswered questions when I last touched this, and I haven't been paying much attention to the changes that have been made to the structure of the app but I know there has been a lot so I suspect I would need to completely rewrite this - which I don't have the inclination to do currently.

Appreciate the feedback and the offer to help, but this isn't something that I'm going to resume in the near future.
